### DISCUSSION TITLE
Including or referencing other justfiles

### DISCUSSION THREAD

**Discussion by tbarusseau** [casey/just#1289] [Ideas]
One feature that I would like is "re-exporting" or "referencing" existing recipes of other `justfile`s.

Considering this directory structure:

```
.
|-- justfile
|-- toy.bin
|   `-- justfile
|-- toy.schemas
|   `-- justfile
|-- toy.script
|   `-- justfile
`-- toy.unity_package
    `-- justfile
```

With this in the top-level `justfile`:

```
@include toy.bin/justfile
@include toy.schemas/justfile
@include toy.script/justfile
@include toy.unity_package/justfile

# Define other recipes if needed
lorem:
    echo "ipsum"

hello:
    echo "world"
```

And this in the lower-level `justfile`s:
```
lorem:
    echo "ipsum"

hello:
    echo "world"
```

Then, we would have this from the top-level directory:

```
$ just
Available recipes:
    # These are the top-level recipes
    lorem
    ipsum

    # These are the lower-level, "included" justfiles' recipes
    @toy.bin:lorem
    @toy.bin:ipsum
    @toy.schemas:lorem
    @toy.schemas:ipsum
    @toy.script:lorem
    @toy.script:ipsum
    @toy.unity_package:lorem
    @toy.unity_package:ipsum
```

Obviously the syntax/keywords are just placeholders. With these changes, `just` doesn't need any context-awareness of any sort, you can just check if the referenced file exists or not and add its existing recipes using a dedicated "nesting" syntax.

It might be a niche usecase, but it would make my life so much better! And might be a dealbreaker whether or not we adopt `just`.

What do you think?

**Comment by runeimp** [comment]
I like the namespacing of includes. But this has been discussed and though namespacing helps in many cases includes are just too simple of an external source mechanism to be reliable with all that Just can do and wants to do in the future. Recipes often need to have a reference to the directory they are in. That context changes with includes. Unfortunately, only a full import system can resolve these sort of context aware issues.
